README.md

```diff
@@ -241,7 +241,58 @@ class EchoHandler implements HandlerInterface
 }
 ```
 
-# Scaling Hermes
+## Graceful shutdown
+
+Hermes worker can be gracefully stopped.
+
+If implementation of `Tomaj\Hermes\Restart\RestartInteface` is provided when initiating `Dispatcher`, Hermes will check `RestartInteface::shouldRestart()` after each processed message. If it returns `true`, Hermes will shutdown _(notice is logged)_.
+
+> **WARNING:** Relaunch is not provided by this library and it should be handled by process controller you use to keep Hermes running _(eg. launchd, daemontools, supervisord, etc.)_.
+
+Currently two methods are implemented.
+
+### SharedFileRestart
+
+Shutdown initiated by touching predefined file.
+
+```php
+$restartFile = '/tmp/hermes_restart';
+$restart = Tomaj\Hermes\Restart\SharedFileRestart($restartFile);
+
+// $log = ...
+// $driver = ....
+$dispatcher = new Dispatcher($driver, $log, $restart);
+```
+
+And then:
+
+```php
+// initiate shutdown; eg. from PHP application
+touch($restartFile, time());
+```
+
+### RedisRestart
+
+Shutdown initiated by storing timestamp to Redis to predefined restart key.
+
+```php
+$redisClient = new Predis\Client();
+$redisRestartKey = 'hermes_restart'; // can be omitted; default value is `hermes_restart`
+$restart = Tomaj\Hermes\Restart\RedisRestart($redisClient, $redisRestartKey);
+
+// $log = ...
+// $driver = ....
+$dispatcher = new Dispatcher($driver, $log, $restart);
+```
+
+And then:
+
+```php
+// initiate shutdown; eg. from PHP application
+$redisClient->set($redisRestartKey, time());
```

Comment:
Just idea - when I see this. Why not add method 'restart()' into `Tomaj\Hermes\Restart\RedisRestart`?

Than I don't need to keep instance of redis client and I just call `$restart->restart()` when i know that I am using this implementation. Maybe it is good idea also to put this method into interface because it can be implemented also in `SharedFileRestart`.

What do you think?

Reply:
it is BC break :)

Reply:
If you are asking why Redis implementation - we need this to restart multiple workers within kubernetes cluster (with shared redis).

---

What would be use of direct `$restart->restart()`? From outside of Hermes I want to use redis/file restart. Not to call directly these implementations of restart interface.

That would probably make sense for some kind of `ManualRestart` implementation?

Maybe I misunderstood your question?

Reply:
Just got it :bulb:  You want to hide that example commands into implementations.

This wouldn't be BC, because we'll add new function. It's just simple way of restarting hermes from application; probably won't be used from inside of Hermes package.

I can add it, let me know your decision.

Reply:
@lulco why it is bc break right now?

Regarding Redis implementation - I like it 👍  I have no doubts about it. 

More I am thinking about this `restart()` method in interface more I like it. If we will have it we can restart all hermes workers from inside of application easily without knowing internals about how restart works or what kind of restart mechanism we are using. Also you can easily introduce command for restarting hermes without knowing how you are dealing with restart.

Can you add it into this PR please? (after that it will be BC break because we will change restart interface)

Reply:
@tomaj not right now, with additional method in interface it would be

Reply:
- Added `RestartInterface::restart()`,
- added it to both implementations,
- fixed related tests (majority of changes),
- rebased current master.